subsys/net/lib/http_server/src/server.c

```diff
@@ -0,0 +1,538 @@
+    if (readBytes < 0) {
+	perror("ERROR reading from socket");
+	close_client_connection(ctx, i);
+    } else if (readBytes == 0) {
+	printf("Connection closed by peer.\n");
+    }
+
+    frame_count = parse_http2_frames(frame, readBytes, frames);
+    print_http2_frames(frames, frame_count);
+
+    size_t content_size = sizeof(content2);
+    sendData(ctx->client_fds[i].fd, content2, content_size, 0x00, 0x01, 0x01);
+
+    readBytes = recv(ctx->client_fds[i].fd, frame, sizeof(frame), 0);
+    if (readBytes < 0) {
+	perror("ERROR reading from socket");
+	close_client_connection(ctx, i);
+    } else if (readBytes == 0) {
+	printf("Connection closed by peer.\n");
+    }
+
+    frame_count = parse_http2_frames(frame, readBytes, frames);
+    print_http2_frames(frames, frame_count);
+
+    hasUpgradeHeader = 1;
+    close_client_connection(ctx, i);
+}
+
+int http2_server_init(struct http2_server_ctx *ctx,
+		      struct http2_server_config *config)
+{
+    /* Create a socket */
+    ctx->sockfd = socket(config->address_family, SOCK_STREAM, 0);
+    if (ctx->sockfd < 0) {
+	perror("socket");
+	exit(EXIT_FAILURE);
+    }
+
+    /* Set up the server address struct according to address family */
+    if (config->address_family == AF_INET) {
+	struct sockaddr_in serv_addr;
+	memset(&serv_addr, 0, sizeof(serv_addr));
+	serv_addr.sin_family = AF_INET;
+	serv_addr.sin_addr.s_addr = INADDR_ANY;
+	serv_addr.sin_port = htons(config->port);
+
+	if (bind(ctx->sockfd, (struct sockaddr *)&serv_addr,
+		 sizeof(serv_addr)) < 0) {
+	    perror("bind");
+	    exit(EXIT_FAILURE);
+	}
+    } else if (config->address_family == AF_INET6) {
+	struct sockaddr_in6 serv_addr;
+	memset(&serv_addr, 0, sizeof(serv_addr));
+	serv_addr.sin6_family = AF_INET6;
+	serv_addr.sin6_addr = in6addr_any;
+	serv_addr.sin6_port = htons(config->port);
+
+	if (bind(ctx->sockfd, (struct sockaddr *)&serv_addr,
+		 sizeof(serv_addr)) < 0) {
+	    perror("bind");
+	    exit(EXIT_FAILURE);
+	}
+    }
+
+    /* Listen for connections */
+    if (listen(ctx->sockfd, MAX_CLIENTS) < 0) {
+	perror("listen");
+	exit(EXIT_FAILURE);
+    }
+
+    /* Initialize client_fds */
+    memset(ctx->client_fds, 0, sizeof(ctx->client_fds));
+    ctx->client_fds[0].fd = ctx->sockfd;
+    ctx->client_fds[0].events = POLLIN;
+    ctx->num_clients = 0;
+
+    return ctx->sockfd;
+}
+
+int http2_server_start(int server_fd, struct http2_server_ctx *ctx)
+{
+    printf("Press 'Ctrl+C' to quit\n");
+    printf("Waiting for incoming connections on port 8080...\n");
+
+    while (1) {
+	int ret = poll(ctx->client_fds, ctx->num_clients + 1, -1);
+	if (ret < 0) {
+	    perror("poll failed");
+	    exit(EXIT_FAILURE);
+	}
+
+	for (int i = 0; i <= ctx->num_clients; i++) {
+	    if (ctx->client_fds[i].revents & POLLIN) {
+		if (i == 0) {
+		    int new_socket = accept_new_client(server_fd);
+
+		    for (int j = 1; j < MAX_CLIENTS; j++) {
+			if (ctx->client_fds[j].fd == 0) {
```

Comment:
I believe I've already mentioned that some time ago, but 0 is still a valid socket descriptor. So in case you want to mark unused `fds` entries, you should use negaive value (most commonly used -1), not 0.